crates/bazed-core/src/buffer/undo_history.rs

```diff
@@ -0,0 +1,225 @@
+use std::collections::BTreeSet;
+
+// Behavior in a sample editing history:
+//
+// initial           => id = 1, history = [0            ], idx = 0, undone = {}
+// edit[new_group=t] => id = 1, history = [0, 1         ], idx = 1, undone = {}
+// edit[new_group=f] => id = 1, history = [0, 1         ], idx = 1, undone = {}
+// edit[new_group=t] => id = 2, history = [0, 1, 2      ], idx = 2, undone = {}
+// edit[new_group=f] => id = 2, history = [0, 1, 2      ], idx = 2, undone = {}
+// edit[new_group=t] => id = 3, history = [0, 1, 2, 3   ], idx = 3, undone = {}
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 2, undone = { 3 }
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 1, undone = { 3, 2 }
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 1, undone = { 3, 2 }
+// redo              => id = 3, history = [0, 1, 2, 3   ], idx = 2, undone = { 3 }
+// edit[new_group=t] => id = 4, history = [0, 1, 2,    4], idx = 3, undone = { 3 }
```

Comment:
These could be module docs or on the struct for easier discoverability.

Reply:
Done!